lib/gpu/gpuProcess.cpp

```diff
@@ -152,23 +155,27 @@ void gpuProcess::main_thread() {
     bool first_run = true;
 
     while (!stop_thread) {
-        int gpu_frame_id = gpu_frame_counter % _gpu_buffer_depth;
+        int icommand = gpu_frame_counter % _gpu_buffer_depth;
+
+        for (auto& command : commands)
+            command[icommand]->start_frame(gpu_frame_counter);
+
         // Wait for all the required preconditions
         // This is things like waiting for the input buffer to have data
         // and for there to be free space in the output buffers.
         // INFO("Waiting on preconditions for GPU[{:d}][{:d}]", gpu_id, gpu_frame_id);
         for (auto& command : commands) {
-            if (command->wait_on_precondition(gpu_frame_id) != 0) {
+            if (command[icommand]->wait_on_precondition() != 0) {
```

Comment:
I think the logic here needs to be adjusted to support HSA, since there we only have one command object per-command.  We could copy the same pointer into each command, or we could abstract it to a function like `queue_commands()` which is different for HSA than for CUDA/OpenCL.